### PR TITLE
HostFactoryResolver - Increase default timeout to thirty seconds

### LINKED ISSUE #60891: System.InvalidOperationException: Unable to build IHost when creating EF Migration with Minimal API dotnet 6
## Description

I have an issue when creating ef migrations for the newest release candidate 6.0.0-rc.2.21480.5 for dotnet 6 with the minimal apis. It seems to be a similar issue as explained here https://github.com/dotnet/aspnetcore/issues/33886

The timeout for creating a WebApplicationFactory<T> at design time seems to be 5 sec and that is not enough for me when I add new migrations, because I need to setup a lot of configuration before. I get this exception when trying to run any command that requires the project to be build like `dotnet ef migrations add AddedPlanningSessions -v` and the DbContext to be fetched from the service provider. I would expect there to be no timeout at all or at least a very large one when using WebApplicationFactory<T> at design time.
```
System.InvalidOperationException: Unable to build IHost
   at Microsoft.Extensions.Hosting.HostFactoryResolver.HostingListener.CreateHost()
   at Microsoft.Extensions.Hosting.HostFactoryResolver.<>c__DisplayClass8_0.<ResolveHostFactory>b__0(String[] args)
   at Microsoft.Extensions.Hosting.HostFactoryResolver.<>c__DisplayClass11_0.<ResolveServiceProviderFactory>b__3(String[] args)
   at Microsoft.EntityFrameworkCore.Design.Internal.AppServiceProviderFactory.CreateFromHosting(String[] args)
```
The issue can be easily reproduced by creating an empty project with minimal api `dotnet new web -o service,` adding a dbcontext to the DI and adding a sleep of 6 sec.

### Include your code

```C#
using Microsoft.EntityFrameworkCore;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddDbContext<ApplicationDbContext>(
    options => options.UseSqlServer("localhost"));

Thread.Sleep(TimeSpan.FromSeconds(6));

var app = builder.Build();

app.MapGet("/", () => "Hello World!");

app.Run();
```

### Include verbose output

Output when building my project where the steps until builder.Build() take too long.

```
dotnet exec --depsfile /Users/ian/Repos/work/planner-service/service/bin/Debug/net6.0/service.deps.json --additionalprobingpath /Users/ian/.nuget/packages --additionalprobingpath /usr/local/share/dotnet/sdk/NuGetFallbackFolder --runtimeconfig /Users/ian/Repos/work/planner-service/service/bin/Debug/net6.0/service.runtimeconfig.json /Users/ian/.dotnet/tools/.store/dotnet-ef/6.0.0-rc.2.21480.5/dotnet-ef/6.0.0-rc.2.21480.5/tools/netcoreapp3.1/any/tools/netcoreapp2.0/any/ef.dll dbcontext info --assembly /Users/ian/Repos/work/planner-service/service/bin/Debug/net6.0/service.dll --project /Users/ian/Repos/work/planner-service/service/service.csproj --startup-assembly /Users/ian/Repos/work/planner-service/service/bin/Debug/net6.0/service.dll --startup-project /Users/ian/Repos/work/planner-service/service/service.csproj --project-dir /Users/ian/Repos/work/planner-service/service/ --root-namespace PlannerService --language C# --framework net6.0 --nullable --working-dir /Users/ian/Repos/work/planner-service/service --verbose
Using assembly 'service'.
Using startup assembly 'service'.
Using application base '/Users/ian/Repos/work/planner-service/service/bin/Debug/net6.0'.
Using working directory '/Users/ian/Repos/work/planner-service/service'.
Using root namespace 'PlannerService'.
Using project directory '/Users/ian/Repos/work/planner-service/service/'.
Remaining arguments: .
Finding DbContext classes...
Finding IDesignTimeDbContextFactory implementations...
Finding application service provider in assembly 'service'...
Finding Microsoft.Extensions.Hosting service provider...
Using environment 'Development'.
Loading config from: global, planner-service.
System.InvalidOperationException: Unable to build IHost
   at Microsoft.Extensions.Hosting.HostFactoryResolver.HostingListener.CreateHost()
   at Microsoft.Extensions.Hosting.HostFactoryResolver.<>c__DisplayClass8_0.<ResolveHostFactory>b__0(String[] args)
   at Microsoft.Extensions.Hosting.HostFactoryResolver.<>c__DisplayClass11_0.<ResolveServiceProviderFactory>b__3(String[] args)
   at Microsoft.EntityFrameworkCore.Design.Internal.AppServiceProviderFactory.CreateFromHosting(String[] args)
An error occurred while accessing the Microsoft.Extensions.Hosting services. Continuing without the application service provider. Error: Unable to build IHost
No application service provider was found.
Finding DbContext classes in the project...
Found DbContext 'PlannerContext'.
Microsoft.EntityFrameworkCore.Design.OperationException: Unable to create an object of type 'PlannerContext'. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728
 ---> System.InvalidOperationException: Unable to resolve service for type 'Microsoft.EntityFrameworkCore.DbContextOptions`1[PlannerService.DAL.PlannerContext]' while attempting to activate 'PlannerService.DAL.PlannerContext'.
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetServiceOrCreateInstance(IServiceProvider provider, Type type)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.<>c__DisplayClass21_4.<FindContextTypes>b__13()
   --- End of inner exception stack trace ---
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.<>c__DisplayClass21_4.<FindContextTypes>b__13()
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(Func`1 factory)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(String contextType)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.GetContextInfo(String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.GetContextInfoImpl(String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.GetContextInfo.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Unable to create an object of type 'PlannerContext'. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728
```

### Include provider and version information

EF Core version: 6.0.0-rc.2.21480.5
Database provider: Microsoft.EntityFrameworkCore.SqlServer 6.0.0-rc.2.21480.5
Target framework: NET 6.0
Operating system:
IDE: VS Code Version: 1.61.1

### PR DESCRIPTION
**Pull request by maryamariyan**
#### Customer Impact

When customer has debugger attached or when using tools like EF, host startup path using WebApplicationBuilder for hosting and WebApplicationFactory<T> for testing, may fail to build and throw an InvalidOperationException with the message “Unable to build IHost.”, because the default timeout is only 5 seconds.

We are proposing to use a higher timeout (1 minute) to be used so that when an application is taking longer to build, we could allow startup code to complete, and the host can start successfully. 

Increasing the timeout should not have a negative impact on experience.

#### TODO
- [x] Add test for this scenario to ensure it works.
- [x] Add environment variable
- [x] improve exception message

#### Risk
Very low.

#### Regression
No, new 6.0 feature/code flow for discovering a host.
Even though this is not a regression (uses WebApplicationBuilder which is new), but since this code is in the ASP.NET Core templates, it regresses the scenario when using default templates.

Fixes https://github.com/dotnet/runtime/issues/60891, https://github.com/dotnet/aspnetcore/issues/33886

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ internal sealed class HostFactoryResolver
         public const string CreateHostBuilder = nameof(CreateHostBuilder);
 
         // The amount of time we wait for the diagnostic source events to fire
-        private static readonly TimeSpan s_defaultWaitTimeout = Debugger.IsAttached ? Timeout.InfiniteTimeSpan : TimeSpan.FromSeconds(5);
+        private static readonly TimeSpan s_defaultWaitTimeout = Debugger.IsAttached ? Timeout.InfiniteTimeSpan : TimeSpan.FromMinutes(1);
```
Comment (halter73) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:25]: Assuming we improve the error message to make it clear that there was a timeout, are we worried that people will just kill the migration before seeing the error if we make this a full minute?

Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs]: changed it to 30 seconds, assuming we would keep the environment variable option.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs [hunk 2]
```diff
@@ -268,7 +284,7 @@ public object CreateHost()
                     // Wait before throwing an exception
                     if (!_hostTcs.Task.Wait(_waitTimeout))
                     {
-                        throw new InvalidOperationException("Unable to build IHost");
+                        throw new InvalidOperationException($"Timed out waiting for the entry point to build the IHost after {s_defaultWaitTimeout}.");
```
Comment (halter73) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:287]: I think we should mention the environment variable if we keep it.

```suggestion
                        throw new InvalidOperationException($"Timed out waiting for the entry point to build the IHost after {s_defaultWaitTimeout}. This timeout can be modified using the '{TimeoutEnvironmentKey}' environment variable.");
```

**Review by halter73 — APPROVED** [review]

**Review by eerhardt — APPROVED** [review]
LGTM

**Comment by davidfowl** [comment]
How did we land on 30 seconds? Just FYI, the reason the keyvault configuration provider tends to be the one that times out is because it makes an HTTP request per key, value pair to get the initial set of configuration values. The default should be like 5 minutes to just cover those scenarios.

**Comment by maryamariyan** [comment]
> How did we land on 30 seconds? 

My thought was, since we are allowing to accept environment variables then it's fair to reduce to 30 seconds. Also because of @halter73 's comment below:

> are we worried that people will just kill the migration before seeing the error if we make this a full minute?

**Comment by davidfowl** [comment]
It should be more than 30 seconds because of the keyvault scenario (if you want some concrete data, look for keyvault configuration provider issues). My vote is 5 minutes

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/HostFactoryResolverTests.cs [hunk 3]
```diff
@@ -253,6 +253,16 @@ public void TopLevelStatements()
             Assert.IsAssignableFrom<IServiceProvider>(factory(Array.Empty<string>()));
         }
 
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
+        public void TopLevelStatementsTestsTimeout()
+        {
+            var assembly = Assembly.Load("TopLevelStatementsTestsTimeout");
+            var factory = HostFactoryResolver.ResolveServiceProviderFactory(assembly, waitTimeout: null); // will use default timeout
```
Comment (halter73) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/HostFactoryResolverTests.cs:260]: Are we okay with this test taking 5 minutes? Would it be okay to test with a non-default timeout?

Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/HostFactoryResolverTests.cs:260]:
```suggestion
            var factory = HostFactoryResolver.ResolveServiceProviderFactory(assembly, s_WaitTimeout);
```

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/TopLevelStatementsTestsTimeout/Program.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+using Microsoft.Extensions.Hosting;
+
+var hostBuilder = new HostBuilder();
+Thread.Sleep(TimeSpan.FromMinutes(6));
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/TopLevelStatementsTestsTimeout/Program.cs:9]:
```suggestion
Thread.Sleep(TimeSpan.FromSeconds(30));
```

**Pull request merged**